pkg/controllermanager/controller/controllerregistration/controllerregistration_control.go

```diff
@@ -0,0 +1,272 @@
+	}
+
+	return c.reconcile(controllerRegistration, logger)
+}
+
+func (c *defaultControllerRegistrationControl) reconcile(controllerRegistration *corev1alpha1.ControllerRegistration, logger logrus.FieldLogger) error {
+	var (
+		err              error
+		result           error
+		installationsMap = map[string]string{}
+
+		mustWriteFinalizer = false
+	)
+
+	seedList, err := c.seedLister.List(labels.Everything())
+	if err != nil {
+		return err
+	}
+
+	for _, seed := range seedList {
+		if seed.DeletionTimestamp == nil {
+			mustWriteFinalizer = true
+		}
+	}
+
+	if mustWriteFinalizer {
+		controllerRegistration, err = kutil.TryUpdateControllerRegistrationWithEqualFunc(c.k8sGardenClient.GardenCore(), retry.DefaultBackoff, controllerRegistration.ObjectMeta, func(c *corev1alpha1.ControllerRegistration) (*corev1alpha1.ControllerRegistration, error) {
+			if finalizers := sets.NewString(c.Finalizers...); !finalizers.Has(FinalizerName) {
+				finalizers.Insert(FinalizerName)
+				c.Finalizers = finalizers.UnsortedList()
+			}
+			return c, nil
+		}, func(cur, updated *corev1alpha1.ControllerRegistration) bool {
+			return sets.NewString(cur.Finalizers...).Has(FinalizerName)
+		})
+		if err != nil {
+			return err
+		}
+	}
+
+	// Live lookup to prevent working on a stale cache and trying to create multiple installations for the same
+	// registration/seed combination.
+	controllerInstallationList, err := c.k8sGardenClient.GardenCore().CoreV1alpha1().ControllerInstallations().List(metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+
+	for _, controllerInstallation := range controllerInstallationList.Items {
+		if controllerInstallation.Spec.RegistrationRef.Name == controllerRegistration.Name {
+			installationsMap[controllerInstallation.Spec.SeedRef.Name] = controllerInstallation.Name
+		}
+	}
+
+	for _, seed := range seedList {
+		if seed.DeletionTimestamp != nil {
+			if installation, ok := installationsMap[seed.Name]; ok {
+				if err := c.k8sGardenClient.GardenCore().CoreV1alpha1().ControllerInstallations().Delete(installation, &metav1.DeleteOptions{}); err != nil && !apierrors.IsNotFound(err) {
+					result = multierror.Append(result, err)
+				}
+			}
+			continue
+		}
+
+		seed, err = kutil.TryUpdateSeedWithEqualFunc(c.k8sGardenClient.Garden(), retry.DefaultBackoff, seed.ObjectMeta, func(s *gardenv1beta1.Seed) (*gardenv1beta1.Seed, error) {
+			if finalizers := sets.NewString(s.Finalizers...); !finalizers.Has(FinalizerName) {
+				finalizers.Insert(FinalizerName)
+				s.Finalizers = finalizers.UnsortedList()
+			}
+			return s, nil
+		}, func(cur, updated *gardenv1beta1.Seed) bool {
+			return sets.NewString(cur.Finalizers...).Has(FinalizerName)
+		})
+		if err != nil {
+			result = multierror.Append(result, err)
+			continue
+		}
+
+		var objectMeta metav1.ObjectMeta
+		if name, ok := installationsMap[seed.Name]; !ok {
+			objectMeta = metav1.ObjectMeta{
+				GenerateName: fmt.Sprintf("%s-", controllerRegistration.Name),
```

Comment:
`objectMeta` with `GenerateName` won't work that nicely with `CreateOrPatch` or does it?